### ISSUE TITLE
Question: Objects inside a list not inferred change runtimeType?

### ISSUE THREAD

**Issue by giorgiofran** [dart-lang/language#1549]
I'm not sure if it is a bug or not, but it is causing me some problems.
Mainly because I need to now which is the right behavior;
Let's start from a case that it seems to me that it is working;
I want to define a List. At present I have three ways:
- defining the expected list type in the instance List => `var fooList = <Object?>[0, 'foo', 10, null];`
- declare the expected type => `List<Object?> fooList = [0, 'foo', 10, null];`
- Let the analyzer infer the type => `var fooList = [0, 'foo', 10, null];`

If I run this simple test 
```dart
print('ListType: ${fooList.runtimeType}');
for (var element in fooList) {
  print('$element, ${element.runtimeType}');
}
```
on all the three cases I have the following result:
```
ListType: List<Object?>
0, int
foo, String
10, int
null, Null
```

And, up to here, we are OK, it work like I'd expected.

No, let's move to a more complex case;

1) Let the system infer the types
```dart
  var fooList = [
        {'first': 4},
        {'second': 'John'},
        {'third': null},
      ];
      print('ListType: ${fooList.runtimeType}');
      for (var element in fooList) {
        print('$element, ${element.runtimeType}');
      }
```
The output is:
```
ListType: List<Map<String, Object?>>
{first: 4}, _InternalLinkedHashMap<String, int>
{second: John}, _InternalLinkedHashMap<String, String>
{third: null}, _InternalLinkedHashMap<String, Null>
```
Nice! This i what I'd expected.
2) define the type on the instance List
```dart
var fooList = <Map<String, Object?>>[
        {'first': 4},
        {'second': 'John'},
        {'third': null},
      ];
      print('ListType: ${fooList.runtimeType}');
      for (var element in fooList) {
        print('$element, ${element.runtimeType}');
      }
```
The output is:
```
ListType: List<Map<String, Object?>>
{first: 4}, _InternalLinkedHashMap<String, Object?>
{second: John}, _InternalLinkedHashMap<String, Object?>
{third: null}, _InternalLinkedHashMap<String, Object?>
```
The element types are all forced to Map<String, Object?>!
This is the weird behavior that is warring me. 
the same happens if I declare the variable type directly:
```dart
List<Map<String, Object?>> fooList = [
        {'first': 4},
        {'second': 'John'},
        {'third': null},
      ];
```
As I said in the beginning, I'm not sure if it is an error, but, if it isn't, I'd need to understand which is the logic behind this behavior.
Thanks in advance.

**Comment by tatumizer** [comment]
Let's start from the "complicated" example:
```dart
var fooList = <Map<String, Object?>>[
        {'first': 4},
        {'second': 'John'},
        {'third': null},
      ];
```
Let's translate it to English:
"Here is my list fooList. It's a list of Maps. Each Map is a map from String to Object?"
What does it mean to be a Map from String to Object? Among other things, it means that you can add an entry to this list.
Let's write it down:
```
fooList[0]["xxx"] = "foo"';
```
Is it supposed to work? Sure. The reason is that you explicitly said: I want this map to hold values of any type (Object? is basically any type). So you get exactly what you wanted, right? If the compiler behaved differently, and internally marked the first entry (in runtime) as a Map from String to int, your assignment `fooList[0]["xxx"] = "foo"';` would fail because of a type mismatch.

When you don't specify the list type explicitly, the compiler "infers" the type as the nearest common superclass of the types of the elements you provided. That explains the difference between the two variants of examples.

Please think about it.

**Comment by giorgiofran** [comment]
> What does it mean to be a Map from String to Object? Among other things, it means that you can add an entry (key, value) to this map, where the value is just anything.
> Let's write it down:
> 
> fooList[0]["xxx"] = "foo"';
> 
> Is it supposed to work? Sure. The reason is that you explicitly said: I want this map to hold values of any type (Object? is basically any type). So you get exactly what you wanted, right? 

Up to here we are OK. As i defined my List in a certain way I can Only add Maps with key String and value anything (Object?)
> If the compiler behaved differently, and internally marked the first entry (in runtime) as a Map from String to int, your assignment fooList[0]["xxx"] = "foo"'; would fail because of a type mismatch.

Here is the point **It should fail** If I wanted my Map `{'intValue' : 4 }` to host any kind of Object I would have defined it as `<String, Object?>{'intValue': 4}.` This way I could add to it any key-value pair with key `String`. 
> 
> When you don't specify the list type explicitly, the compiler "infers" the type as the nearest common superclass of the types of the elements you provided. The same happens with map literals inside the list. That explains the difference between the two variants of examples.

So definitely, the problem is how the compiler infers the type. By default, i.e., if the user does not provide an explicit definition like<String, int> for example, from my point of view it should keep the one it would have inferred normally if it wouldn't have been the element of another collection, while you say that it considers the "parent" definition" (if any) to set the type to the nearest common superclass.

From my point of view it is a bug:
consider this:
```dart
var list = ['foo'];
```
In this case list is of type `List<String>`
So, the compiler, when encountered a ['foo'] element recognized it as a `List<String>`.
now the same list inside another List:
```dart
var list = <Object>[['foo']];
```
Here the list is of type `List<Object>` _correct_, and the element is of type `List<String>` _correct_!
Also here the compiler, when encountered a ['foo'] element recognized it as a `List<String>`.

Now this:
```dart
var list = <List<Object>>[['foo']];
```
here the list is of type `List<List<Object>>` _correct_, and the element is of type `List<Object>` !!! Why?
Here, instead, the compiler, when encountered a ['foo'] element didn't recognized it as a `List<String>`.
The compiler should infer types always the same way, otherwise it is a mess.

**Comment by tatumizer** [comment]
I see your point. But the opposite point also has some merit. Let's consider some examples:
```dart
  var list1 = <List>[[5]];
  print(list1[0].runtimeType); // List<dynamic>
  var list2 = <List<Object>>[[5]];
  print(list2[0].runtimeType); // List<Object>
  var list3 = [[5]];
  print(list3[0].runtimeType); // List<int>
```
My theory was basically that "explicit beats implicit"; even though "explicit" comes from a parent, it still carries some flavor of explicitness. Now it's a question of degree - it is certainly not 100% explicit, but maybe 50% explicit, so why the compiler decided in favor of these 50% and not the other 50% (which is implicit). After thinking deeply about the phenomenon, I came up with another theory: explicitness is a dominant trait, so the child inherits it unconditionally.

It may seem that list1 provides a counterexample to my theory: here, we don't explicitly specify the type of elements, but the compiler does it for us: in the absence of direct indication of user's choice for the parameter type, the compiler by default sets it to "dynamic". It's very difficult to argue that our definition was explicit. Difficult, but possible! Even 1% of explicitness beats implicitness: remember, it's a dominant trait!

If we look at the examples from this angle, the treatment of our definitions looks pretty consistent.

 What's really missing from the language is the ability for the user to explicitly say: "Don't get me wrong, I'm NOT explicit here!"
That is, the language could hypothetically support declarations like
`var list1 = <List<inferred>>[[5]];` - and I vaguely remember seeing some "issue" requesting just that (maybe in the form `List<?>` or something).

I'm not associated with dart team in any manner, so the best I can do is speculate. Let's wait for someone more competent to respond.

**Comment by giorgiofran** [comment]
@tatumizer thanks, your point of view is valid too. That language feature is quite interesting :-)
Anyway, I hope that someone from the team could give as a proper response.

**Comment by eernstg** [comment]
@giorgiofran, note that @tatumizer writes a lot of proper responses. ;-) But I can add some words about the given examples, with details given [here](https://github.com/dart-lang/language/blob/3500b68861bca9200339921dac35d8d8f55d1873/specification/dartLangSpec.tex#L9612), PDF [here](https://spec.dart.dev/DartLangSpecDraft.pdf):

```dart
var list = ['foo'];
```

> In this case list is of type `List<String>`

This is because there is no context type, so the type argument of the list literal is computed based on the contents (the least upper bound of the types of the elements), which yields `String`.

> So, the compiler, when encountered a `['foo']` element recognized it as a `List<String>`.

That would not be true if there is a non-trivial context type. For instance:

```dart
void main () {
  List<Object> xs = ['foo']; // Inferred as `<Object>['foo`]`.
  xs.add(Object()); // Works! If we had inferred `<String>['foo']`, this would throw.
}
```

> now the same list inside another `List`:

```dart
var list = <Object>[['foo']];
```

> Here the list is of type `List<Object>` correct,

The outer list has that type because it was specified explicitly, no inference here.

> and the element is of type `List<String>` correct!

The context type for the inner list is `Object`, which doesn't constrain the type argument of the inner list. So we use bottom up inference again and obtain a `<String>['foo']`.

> Also here the compiler, when encountered a `['foo']` element recognized it as a `List<String>`.

Based on the context type.

> Now this:

```dart
var list = <List<Object>>[['foo']];
```

Here, you've specified that the outer list must be a `List<List<Object>>` (again, no inference occurs for the outer list, the type argument is given). This means that the context type for the inner list is `List<Object>`, and we get `<Object>['foo']`.

**Comment by eernstg** [comment]
If lists had been soundly covariant (in particular, if they were immutable) then it could have made sense to use bottom up inference consistently. Lists could then also be declared to be invariant, in which case there wouldn't be this choice of type argument in many cases (if you wish to pass `[1]` as an actual argument where the formal parameter has type `List<num>` then it _must_ become `<num>[1]` because `<int>[1]` would be a compile-time error). We may add [sound variance](https://github.com/dart-lang/language/issues/524) to Dart, but it's not likely to be acceptable to change lists to be invariant or immutable, that would be a massively breaking change.

This means that we will work with lists where the covariance is checked dynamically for a while. And that again motivates the example that I gave:

```dart
void main () {
  List<Object> xs = ['foo']; // Inferred as `<Object>['foo`]`, because the context type wins.
  xs.add(Object()); // Works! If we had inferred `<String>['foo']`, this would throw.
}
```

If we had insisted on inferring `['foo']` bottom up and gotten `<String>['foo']` then the initialization of `xs` would have worked just fine, but any attempt to add a non-string to `xs` would throw. That's not likely to be the desired behavior with the declarations as shown.

@giorgiofran, I think it would make sense if you closed this issue because it is all working as intended (I haven't seen anything to the contrary), and because (I think) there are no concrete issues or proposals here. But you're of course welcome to create new issues about anything and everything odd or wrong that you encounter in Dart!
